tests/robustness/validate/patch_history_test.go

```diff
@@ -0,0 +1,227 @@
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendDelete("key", start, stop, &clientv3.DeleteResponse{}, nil)
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed delete remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendDelete("key", start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: true,
+		},
+		{
+			name: "successful txn remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{}, []clientv3.Op{}, start, stop, &clientv3.TxnResponse{}, nil)
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed txn remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{}, []clientv3.Op{}, start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: false,
+		},
+		{
+			name: "successful txn put/delete remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{clientv3.OpPut("key", "value")}, []clientv3.Op{clientv3.OpDelete("key")}, start, stop, &clientv3.TxnResponse{}, nil)
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed txn put/delete remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{clientv3.OpPut("key", "value")}, []clientv3.Op{clientv3.OpDelete("key")}, start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed txn delete/put remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{clientv3.OpDelete("key")}, []clientv3.Op{clientv3.OpPut("key", "value")}, start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed txn put&delete/put&delete is dropped",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{clientv3.OpPut("key", "value1"), clientv3.OpDelete("key")}, []clientv3.Op{clientv3.OpPut("key", "value2"), clientv3.OpDelete("key")}, start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: false,
+		},
```

Comment:
For other reviewers, that's the most important one matched the above statement. 

> Currently etcd nor Kubernetes traffic code doesn't generate such transactions.

Reply:
Not accurate, this also applies to `failed txn put/delete remains`, `failed txn delete/put remains` and many other 
test cases.